lte/gateway/c/core/oai/include/map.h

```diff
@@ -0,0 +1,232 @@
+
+    case MAP_BAD_PARAMETER_VALUE:
+      return "MAP_BAD_PARAMETER_VALUE";
+      break;
+
+    case MAP_EMPTY:
+      return "MAP_EMPTY";
+      break;
+
+    case MAP_DUMP_FAIL:
+      return "MAP_DUMP_FAIL";
+      break;
+
+    default:
+      return "UNKNOWN map_rc_t";
+  }
+}
+
+/***************************************************************************
+**                                                                        **
+** Name:    map_ts_s                                                      **
+**                                                                        **
+** Description: This is a generic structure for maps.It is implemented    **
+**              using template struct definitions.                        **
+**                                                                        **
+** Parameters:  keyT     - data type of key                               **
+**              valueT   - data type of value                             **
+**              Hash     - used to pass a custom hash function.           **
+**                         Default: std::hash<keyT>                       **
+**              KeyEqual - it is used to overload the EqualTo operator.   **
+**                         When using a userdefined struct as key,        **
+**                         pass a suitable EqualTo method.                **
+**                            Default: std::equal_to<keyT>                **
+**                                                                        **
+** APIs:       set_name() , get_name(), get(), insert(), delete()        **
+**                                                                        **
+***************************************************************************/
+template<
+    typename keyT, typename valueT, class Hash = std::hash<keyT>,
+    class KeyEqual = std::equal_to<keyT>>
+struct map_ts_s {
+  std::mutex mutex_obj;
+  std::unordered_map<keyT, valueT, Hash, KeyEqual> umap;
+  std::string name;
+  bool log_enabled;
+
+  void set_name(std::string umap_name) { name = umap_name; }
+  std::string get_name() { return name; }
+
+  /***************************************************************************
+  **                                                                        **
+  ** Name:    get()                                                         **
+  **                                                                        **
+  ** Description: Takes key and valueP as parameters.If the key exists,     **
+  **              corresponding value is returned through the valueP,       **
+  **              else returns error.                                       **
+  **                                                                        **
+  ***************************************************************************/
+  map_rc_t get(const keyT key, valueT* valueP) {
+    if (umap.empty()) {
+      return MAP_EMPTY;
+    }
+    if (valueP == nullptr) {
```

Comment:
this condition could be hit if the `valueP` pointer passed is not initialized, why do we return `MAP_BAD_PARAMETER_VALUE` here?

Reply:
Yes, this condition checks if the valueP is initialized. This is needed as, after fetching the key-value pair we are coping the fetched value into *valueP. If not, can you please suggest a better way of doing this?

Reply:
but this is done _before_ fetching the key-value pair no? or am i missing something else?

Reply:
Agree. Before fetching the caller should provide valid key-value parameters.
this was a safety check if the caller passes valueP value nullptr unintentionally. below piece code will not crash.
```
    if (search_result != umap.end()) {
      *valueP = search_result->second;
      return MAP_OK;
      }
  ```
 it was our assumptions based on existing hash table validation checks.
 @ardzoht,  let's conclude from your side. we will address this accordingly.